ONSdigital/research-and-development
# 498 postcode in json block

## Pull Request submission

This pull request allows users to run the pipeline much faster because one of the slowest steps (postcode validation) is now inside the block of code which run when loading Json - and this only needs to be run once.

Once validation has been done in the json block the validated dataset is written out to feather and this can be optionally loaded next time the pipeline is run. Loading data from a feather should be about 10-20 times faster and because it skips all the validation and datatype coersion steps, the pipeline runs end to end about 5 times faster. 

There are also many other improvements to the code. These implementing include things like error handling and correcting the logic in the `check_data_shape` function in validation. 

I have also tided a lot of the code up in `staging_main.py` by creating several functions that take care of a series of related tasks and calling them within `run_staging`. This simplifies the previously complicated logic in `run_staging` however it still does not pass flake8 because it remains too complicated. 

Things that could be improved:
- Further tidying up of `staging_main.py` to make it leaner - moving functions out into sub-modules.
- Loading all the mappers from feather optionally - ideally with a the same function for each mapper that takes a "load_mapper_feather" parameter and checks if the feather exists, and if not, loads from csv. 


*Instructions to test*

- Alter the config settings and load from json (go and make a coffee because this takes ages)
    - Check that a feather is written out
- Optionally delete or move the feather so this is forced to happen, whatever the config settings
- Alter the config setting again to load from feather
- Check that the pipeline runs from end to end in both cases (json and feather loading)
- Run the pipeline on `develop` and check that the outputs are identical on this branch


#### Closes or fixes

RDRP-498


#### Code

- [x] **Code runs** The code runs on my machine and/or CDSW
- [x] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [x] **Requirements** My/our code functions according to the requirements of the ticket
- [x] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [x] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
    - [x] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [x] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [x] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)


#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.
